### PR TITLE
Address Xiangrui's comments.

### PR DESCRIPTION
**Pull request by andrewmchen**

### CODE REVIEW

File: databricks_cli/jobs/cli.py [hunk 1]
```diff
@@ -111,19 +115,20 @@ def list_cli(output):
         click.echo(tabulate(_jobs_to_table(jobs_json), tablefmt='plain'))
 
 
-@click.command(context_settings=CONTEXT_SETTINGS)
-@click.option('--job-id', required=True, type=int)
+@click.command(context_settings=CONTEXT_SETTINGS,
+               short_help="Delete the specified job from the Databricks Job Service.")
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: * `Deletes`
* Remove `from the Databricks Job Service`. The context is sufficient.

File: databricks_cli/jobs/cli.py [hunk 2]
```diff
@@ -111,19 +115,20 @@ def list_cli(output):
...
+@click.option('--job-id', required=True, type=JobIdClickType())
 @require_config
 @eat_exceptions
 def delete_cli(job_id):
     """
-    Delete the specified job from the Databricks Job Service
+    Delete the specified job from the Databricks Job Service.
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: Ditto.

File: databricks_cli/jobs/cli.py [hunk 3]
```diff
@@ -156,14 +166,18 @@ def run_now_cli(job_id, jar_params, notebook_params, python_params, spark_submit
...
-              expose_value=False, is_eager=True)
+              expose_value=False, is_eager=True, help=version)
 @require_config
 @eat_exceptions
 def jobs_group():
     """
     Utility to interact with the Jobs service.
+
+    This is a wrapper around the jobs API (https://docs.databricks.com/api/latest/jobs.html).
+    /jobs/runs API endpoints are handled by ``databricks runs``.
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: `Job runs are handled by ...`

File: databricks_cli/clusters/cli.py [hunk 4]
```diff
@@ -24,37 +24,40 @@
...
+@click.option('--json-file', default=None, type=click.Path(),
+              help='File containing JSON request to POST to /api/2.0/clusters/create.')
+@click.option('--json', default=None, type=JsonClickType(),
+              help=JsonClickType.help)
 @require_config
 @eat_exceptions
 def create_cli(json_file, json):
     """
-    Creates a Databricks Cluster
+    Creates a cluster in Databricks.
```
Comment (mengxr) [inline databricks_cli/clusters/cli.py]: `Creates a Databricks cluster`

File: databricks_cli/clusters/cli.py [hunk 5]
```diff
@@ -157,17 +172,18 @@ def list_node_types_cli():
 @eat_exceptions
 def spark_versions_cli():
     """
-    List possible spark versions for a cluster.
+    Lists possible spark versions for a cluster.
```
Comment (mengxr) [inline databricks_cli/clusters/cli.py]: We should say `Apache Spark or Databricks Runtime versions`. Let's discuss this offline.

File: databricks_cli/clusters/cli.py [hunk 6]
```diff
@@ -157,17 +172,18 @@ def list_node_types_cli():
...
 
     The output format is specified in
     https://docs.databricks.com/api/latest/clusters.html#spark-versions
     """
     click.echo(pretty_format(spark_versions()))
 
 
-@click.group(context_settings=CONTEXT_SETTINGS)
+@click.group(context_settings=CONTEXT_SETTINGS,
+             short_help='Utility to interact with the Clusters service.')
```
Comment (mengxr) [inline databricks_cli/clusters/cli.py]: `with Databricks clusters`

File: databricks_cli/clusters/cli.py [hunk 7]
```diff
@@ -24,37 +24,40 @@
...
-@click.command(context_settings=CONTEXT_SETTINGS)
-@click.option('--cluster-id', required=True)
+@click.command(context_settings=CONTEXT_SETTINGS,
+               short_help='Starts a terminated Databricks cluster given its id.')
+@click.option('--cluster-id', required=True, type=ClusterIdClickType())
 @require_config
 @eat_exceptions
 def start_cli(cluster_id):
     """
-    Starts a terminated Spark cluster given its id.
```
Comment (mengxr) [inline databricks_cli/clusters/cli.py]: Be consistent with `id` capitalization. I prefer `ID` in all places.

File: databricks_cli/click_types.py [hunk 8]
```diff
@@ -25,18 +25,37 @@
...
 
     @classmethod
     def is_table(cls, value):
-        return value == 'table'
+        return value is not None and value.lower() == 'table'
+
+
+class JsonClickType(ParamType):
+    name = 'JSON'
+    help = 'Valid JSON string. Cannot be provided with --json-file.'
```
Comment (mengxr) [inline databricks_cli/click_types.py]: * `JSON string to POST to /api/2.0/jobs/create`.
* A little hard to understand "cannot be provided with`. I think we should say requiring either one in both help messages.

File: databricks_cli/click_types.py [hunk 9]
```diff
@@ -25,18 +25,37 @@
...
+
+
+class JobIdClickType(ParamType):
+    name = 'JOB_ID'
```
Comment (mengxr) [inline databricks_cli/click_types.py]: No help message. It is useful to tell users where to find it in UI.

File: databricks_cli/click_types.py [hunk 10]
```diff
@@ -25,18 +25,37 @@
...
+
+
+class RunIdClickType(ParamType):
+    name = 'RUN_ID'
```
Comment (mengxr) [inline databricks_cli/click_types.py]: No help message. Where to find it in UI? (Seems we need to add it first.)

File: databricks_cli/click_types.py [hunk 11]
```diff
@@ -25,18 +25,37 @@
...
+
+
+class ClusterIdClickType(ParamType):
+    name = 'CLUSTER_ID'
```
Comment (mengxr) [inline databricks_cli/click_types.py]: No help message. Same, where to find it in UI?

File: databricks_cli/jobs/cli.py [hunk 12]
```diff
@@ -84,23 +87,24 @@ def reset_cli(json_file, json, job_id):
...
 @eat_exceptions
 def list_cli(output):
     """
-    Lists the jobs in the Databricks Job Service
+    Lists the jobs in the Databricks Job Service.
 
     By default the output format will be a human readable table with the following fields
 
-      - job_id
+      - Job ID
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: `job ID` (minor)

File: databricks_cli/jobs/cli.py [hunk 13]
```diff
@@ -84,23 +87,24 @@ def reset_cli(json_file, json, job_id):
...
 
-      - settings.name
+      - Job name
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: `job name` (minor)

File: databricks_cli/jobs/cli.py [hunk 14]
```diff
@@ -84,23 +87,24 @@ def reset_cli(json_file, json, job_id):
...
 
     A json formatted output can also be requested by setting the --output parameter to "json"
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: `JSON`

File: databricks_cli/jobs/cli.py [hunk 15]
```diff
@@ -50,10 +52,11 @@ def create_cli(json_file, json):
 
 
 @click.command(context_settings=CONTEXT_SETTINGS)
-@click.option('--json-file', default=None,
-              help='File containing json to POST to /jobs/reset.')
-@click.option('--json', default=None)
-@click.option('--job-id', required=True, help='The job_id to reset')
+@click.option('--json-file', default=None, type=click.Path(),
+              help='File containing JSON request to POST to /api/2.0/jobs/reset.')
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: This is not correct.

File: databricks_cli/jobs/cli.py [hunk 16]
```diff
@@ -50,10 +52,11 @@ def create_cli(json_file, json):
...
+@click.option('--json', default=None, type=JsonClickType(),
+              help=JsonClickType.help)
+@click.option('--job-id', required=True, type=JobIdClickType(), help='The job_id to reset')
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: `job ID`

File: databricks_cli/jobs/cli.py [hunk 17]
```diff
@@ -63,9 +66,9 @@ def reset_cli(json_file, json, job_id):
     The specification for the json option can be found
     https://docs.databricks.com/api/latest/jobs.html#jobsjobsettings
 
-    NOTE. The json parameter describe above is not the same as what is normally POSTed
+    NOTE. The json parameter described above is not the same as what is normally POSTed
     in the request body to the reset endpoint. Instead it is the object
-    defined in the top level "new_settings" field. The job_id is provided
+    defined in the top level "new_settings" field. The job ID is provided
     by the job-id option.
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: `--job-id` (minor)

File: databricks_cli/jobs/cli.py [hunk 18]
```diff
@@ -134,11 +139,16 @@ def get_cli(job_id):
...
 
 @click.command(context_settings=CONTEXT_SETTINGS)
-@click.option('--job-id', required=True, type=int)
-@click.option('--jar-params', default=None)
-@click.option('--notebook-params', default=None)
-@click.option('--python-params', default=None)
-@click.option('--spark-submit-params', default=None)
+@click.option('--job-id', required=True, type=JobIdClickType())
+@click.option('--jar-params', default=None, type=JsonClickType(),
+              help='JSON array of parameters. i.e. ["param1", "param2"]')
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: `JSON string specifying an array of parameters`. It is not clear if users need to quote the entire value if we say `JSON array`.

File: databricks_cli/jobs/cli.py [hunk 19]
```diff
@@ -134,11 +139,16 @@ def get_cli(job_id):
...
+@click.option('--notebook-params', default=None, type=JsonClickType(),
+              help='JSON map of key-value pairs. i.e. {"name": "john doe", "age": 35}')
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: Ditto.

File: databricks_cli/jobs/cli.py [hunk 20]
```diff
@@ -134,11 +139,16 @@ def get_cli(job_id):
...
+@click.option('--python-params', default=None, type=JsonClickType(),
+              help='JSON array of parameters. i.e. ["param1", "param2"]')
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: Ditto.

File: databricks_cli/jobs/cli.py [hunk 21]
```diff
@@ -134,11 +139,16 @@ def get_cli(job_id):
...
+@click.option('--spark-submit-params', default=None, type=JsonClickType(),
+              help='JSON array of parameters. i.e. '
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: Ditto.

File: databricks_cli/jobs/cli.py [hunk 22]
```diff
@@ -134,11 +139,16 @@ def get_cli(job_id):
...
-@click.option('--python-params', default=None)
-@click.option('--spark-submit-params', default=None)
+@click.option('--job-id', required=True, type=JobIdClickType())
+@click.option('--jar-params', default=None, type=JsonClickType(),
+              help='JSON array of parameters. i.e. ["param1", "param2"]')
+@click.option('--notebook-params', default=None, type=JsonClickType(),
+              help='JSON map of key-value pairs. i.e. {"name": "john doe", "age": 35}')
+@click.option('--python-params', default=None, type=JsonClickType(),
+              help='JSON array of parameters. i.e. ["param1", "param2"]')
+@click.option('--spark-submit-params', default=None, type=JsonClickType(),
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: Do we support this? It is not in the API doc.

File: databricks_cli/jobs/cli.py [hunk 23]
```diff
@@ -134,11 +139,16 @@ def get_cli(job_id):
...
-@click.option('--jar-params', default=None)
-@click.option('--notebook-params', default=None)
-@click.option('--python-params', default=None)
-@click.option('--spark-submit-params', default=None)
+@click.option('--job-id', required=True, type=JobIdClickType())
+@click.option('--jar-params', default=None, type=JsonClickType(),
+              help='JSON array of parameters. i.e. ["param1", "param2"]')
+@click.option('--notebook-params', default=None, type=JsonClickType(),
+              help='JSON map of key-value pairs. i.e. {"name": "john doe", "age": 35}')
+@click.option('--python-params', default=None, type=JsonClickType(),
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: Do we support this? It is not in the API doc.

File: databricks_cli/clusters/cli.py [hunk 24]
```diff
@@ -118,6 +124,14 @@ def list_cli(output):
 
     Returns information about all currently active clusters, and up
     to 100 most recently terminated clusters in the past 7 days.
+
+    By default the output format will be a human readable table with the following fields
+
+      - Cluster ID
```
Comment (mengxr) [inline databricks_cli/clusters/cli.py]: (minor) cluster ID

File: databricks_cli/runs/cli.py [hunk 25]
```diff
@@ -65,23 +67,30 @@ def _runs_to_table(runs_json):
...
+      - Run ID
 
-    By default
+      - Run name
 
-      - limit is set to 20. This means 20 runs are listed. limit must be in between 0 to 1000
+      - Life cycle state
 
-      - offset is set to 0. The offset is relative to the most recent run_id.
+      - Result state
```
Comment (mengxr) [inline databricks_cli/runs/cli.py]: Mention `n/a`

File: databricks_cli/runs/cli.py [hunk 26]
```diff
@@ -65,23 +67,30 @@ def _runs_to_table(runs_json):
...
-@eat_exceptions
+@eat_exceptions # noqa
 def list_cli(job_id, active_only, completed_only, offset, limit, output): # noqa
     """
-    Lists runs from the Databricks Job Scheduler.
+    Lists job runs.
 
     The limit and offset determine which runs will be listed. Runs are always listed
-    by descending order of run_id.
+    by descending order of run start time or run_id.
```
Comment (mengxr) [inline databricks_cli/runs/cli.py]: `or run_id` -> `and run ID`.

File: databricks_cli/runs/cli.py [hunk 27]
```diff
@@ -65,23 +67,30 @@ def _runs_to_table(runs_json):
...
-@click.option('--output', default=None, help=OutputClickType.help, type=OutputClickType())
+@click.option('--offset', default=None, type=int,
+              help='Set to 0 by default. The offset is relative to the most recent run ID.')
+@click.option('--limit', default=None, type=int,
+              help='Set to 20 runs by default. The limit determines the number of runs listed. '
+                   'Limit must be between 0 and 1000.')
+@click.option('--output', help=OutputClickType.help, type=OutputClickType())
 @require_config
-@eat_exceptions
+@eat_exceptions # noqa
```
Comment (mengxr) [inline databricks_cli/runs/cli.py]: What does `noqa` mean?

File: databricks_cli/runs/cli.py [hunk 28]
```diff
@@ -65,23 +67,30 @@ def _runs_to_table(runs_json):
               help='If specified, only active runs will be listed')
 @click.option('--completed-only', is_flag=True, default=None,
               help='If specifed, only completed runs will be listed')
```
Comment (mengxr) [inline databricks_cli/runs/cli.py]: typo `specified`

File: databricks_cli/runs/cli.py [hunk 29]
```diff
@@ -65,23 +67,30 @@ def _runs_to_table(runs_json):
...
-@click.option('--offset', default=None, type=int)
-@click.option('--limit', default=None, type=int)
-@click.option('--output', default=None, help=OutputClickType.help, type=OutputClickType())
+@click.option('--offset', default=None, type=int,
+              help='Set to 0 by default. The offset is relative to the most recent run ID.')
```
Comment (mengxr) [inline databricks_cli/runs/cli.py]: Explain the param first and then mention the default value.

File: databricks_cli/runs/cli.py [hunk 30]
```diff
@@ -65,23 +67,30 @@ def _runs_to_table(runs_json):
...
+@click.option('--limit', default=None, type=int,
+              help='Set to 20 runs by default. The limit determines the number of runs listed. '
```
Comment (mengxr) [inline databricks_cli/runs/cli.py]: Ditto.

File: databricks_cli/clusters/cli.py (hunk 24 above)
Comment (andrewmchen) [reply databricks_cli/clusters/cli.py]: Hm, what's the reason for this?

File: databricks_cli/jobs/cli.py (hunk 23 above)
Comment (andrewmchen) [reply databricks_cli/jobs/cli.py]: It is supported and we'll add it to the API doc this release.

File: databricks_cli/jobs/cli.py (hunk 22 above)
Comment (andrewmchen) [reply databricks_cli/jobs/cli.py]: It is supported and we'll add it to the API doc this release.

File: databricks_cli/runs/cli.py (hunk 27 above)
Comment (andrewmchen) [reply databricks_cli/runs/cli.py]: Tells the linter to avoid this line. For some reason, I was getting a "too many parameters" lint on it otherwise.

File: databricks_cli/jobs/cli.py [hunk 31]
```diff
@@ -84,25 +89,26 @@ def reset_cli(json_file, json, job_id):
...
     By default the output format will be a human readable table with the following fields
 
-      - job_id
+      - Job ID
 
-      - settings.name
+      - Job name
 
-    A json formatted output can also be requested by setting the --output parameter to "json"
+    A JSON formatted output can also be requested by setting the --output parameter to "json"
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: `json` -> `JSON`

File: databricks_cli/jobs/cli.py [hunk 32]
```diff
@@ -136,11 +143,17 @@ def get_cli(job_id):
...
-@click.option('--job-id', required=True, type=int)
-@click.option('--jar-params', default=None)
-@click.option('--notebook-params', default=None)
-@click.option('--python-params', default=None)
-@click.option('--spark-submit-params', default=None)
+@click.option('--job-id', required=True, type=JobIdClickType(), help=JobIdClickType.help)
+@click.option('--jar-params', default=None, type=JsonClickType(),
+              help='JSON string specifying an array of parameters. i.e. ["param1", "param2"]')
+@click.option('--notebook-params', default=None, type=JsonClickType(),
+              help='JSON string specifying map of key-value pairs. '
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: `a map`

File: databricks_cli/jobs/cli.py [hunk 33]
```diff
@@ -136,11 +143,17 @@ def get_cli(job_id):
...
+                   'i.e. {"name": "john doe", "age": 35}')
+@click.option('--python-params', default=None, type=JsonClickType(),
+              help='JSON string specifying array of parameters. i.e. ["param1", "param2"]')
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: `an array`

File: databricks_cli/jobs/cli.py [hunk 34]
```diff
@@ -136,11 +143,17 @@ def get_cli(job_id):
...
+@click.option('--spark-submit-params', default=None, type=JsonClickType(),
+              help='JSON string specifying array of parameters. i.e. '
```
Comment (mengxr) [inline databricks_cli/jobs/cli.py]: `an array`

File: databricks_cli/runs/cli.py [hunk 35]
```diff
@@ -52,10 +54,11 @@ def _runs_to_table(runs_json):
     for r in runs_json.get('runs', []):
         run_id = r.get('run_id', 'no_run_id')
         run_name = r.get('run_name', 'no_run_name')
-        life_cycle_state = r.get('state', {}).get('life_cycle_state', 'no_life_cycle_state')
-        result_state = r.get('state', {}).get('result_state', 'no_result_state')
-        run_page_url = r.get('run_page_url', 'no_run_page_url')
-        ret.append((run_id, run_name, life_cycle_state, result_state, run_page_url))
+        life_cycle_state = r.get('state', {}).get('life_cycle_state', 'n/a')
+        result_state = r.get('state', {}).get('result_state', 'n/a')
+        run_page_url = r.get('run_page_url', 'n/a')
```
Comment (mengxr) [inline databricks_cli/runs/cli.py]: just realized that we don't output run page url for normal job runs. maybe we should return the run page url as well (not part of this PR)

File: databricks_cli/clusters/cli.py (hunk 24 above)
Comment (mengxr) [reply databricks_cli/clusters/cli.py]: nvm, this is just style. all we need is consistency.
